synapse/http/request_metrics.py

```diff
@@ -98,14 +98,76 @@
     "response_size", labels=["method", "servlet", "tag"]
 )
 
+# In flight metrics are incremented while the requests are in flight, rather
+# than when the response was written.
+
+in_flight_requests_ru_utime = metrics.register_counter(
+    "in_flight_requests_ru_utime_seconds", labels=["method", "servlet"],
+)
+
+in_flight_requests_ru_stime = metrics.register_counter(
+    "in_flight_requests_ru_stime_seconds", labels=["method", "servlet"],
+)
+
+in_flight_requests_db_txn_count = metrics.register_counter(
+    "in_flight_requests_db_txn_count", labels=["method", "servlet"],
+)
+
+# seconds spent waiting for db txns, excluding scheduling time, when processing
+# this request
+in_flight_requests_db_txn_duration = metrics.register_counter(
+    "in_flight_requests_db_txn_duration_seconds", labels=["method", "servlet"],
+)
+
+# seconds spent waiting for a db connection, when processing this request
+in_flight_requests_db_sched_duration = metrics.register_counter(
+    "in_flight_requests_db_sched_duration_seconds", labels=["method", "servlet"]
+)
+
+_in_flight_requests_count = metrics.register_gauge(
+    "in_flight_requests_count", labels=["method", "servlet"]
+)
+
+
+# The set of all in flight requests.
+_in_flight_requests = set()
+
+
+def _collect_in_flight():
+    """Called just before metrics are collected, so we use it to update all
+    the in flight request metrics
+    """
+
+    # Map from (method, name) -> int, the number of in flight requests of that
+    # type
+    counts = {}
+
+    for rm in _in_flight_requests:
+        rm.update_metrics()
+        key = (rm.method, rm.name,)
+        counts[key] = counts.get(key, 0) + 1
+
+    for (method, name), count in counts.iteritems():
+        _in_flight_requests_count.set(count, method, name)
```

Comment:
will this not end up with stale results, because we never properly clear it?

Reply:
This gets run before every call to `/metrics`, so should be fine. OTOH we may as well make it clearer by using a callback metrics

Reply:
yes, but we'll still end up with stale items in `_in_flight_requests_count` ?

+1 to fixing with a callbackmetrics